programs/bpf_loader/src/syscalls.rs

```diff
@@ -2299,126 +2516,1803 @@ fn call<'a>(
+                self.loader_id,
+                true
+            ),
+            result
+        );
+        // Handle expected result > MAX_BIGNUM_BYTE_SIZE bytes
+        if bn_syscall_result.data_len > MAX_BIGNUM_BYTE_SIZE {
+            *result = Err(SyscallError::BigNumberResultSizeError.into());
+        }
+        // Handle buffer too small or empty strings
+        else if bn_syscall_result.data_len == 0 || raw_string.is_empty() {
+            *result = Err(SyscallError::BigNumberToBytesError.into())
+        } else {
+            // Process BigNum call and populate results
+            match BigNum::from_dec_str(raw_string) {
+                Ok(bn) => bignum_to_ffibignumber(
+                    &bn,
+                    bn_syscall_result,
+                    self.loader_id,
+                    memory_mapping,
+                    result,
+                ),
+                Err(_) => *result = Err(SyscallError::BigNumberFromDecStrError.into()),
+            };
+        }
+    }
+}
+
+/// BIGNUM sol_bignum_sqr
+struct SyscallBigNumFromSlice<'a> {
+    cost: u64,
+    word_cost: u64,
+    word_div_cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
+    loader_id: &'a Pubkey,
+}
+impl<'a> SyscallObject<BpfError> for SyscallBigNumFromSlice<'a> {
+    fn call(
+        &mut self,
+        rhs_ffi_in_addr: u64,
+        bn_ffi_out_addr: u64,
+        _arg3: u64,
+        _arg4: u64,
+        _arg5: u64,
+        memory_mapping: &MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        // Get the RHS FfiBigNumber value structure
+        let bn_syscall_rhs = question_mark!(
+            translate_type::<FfiBigNumber>(memory_mapping, rhs_ffi_in_addr, self.loader_id, true),
+            result
+        );
+        // Compute costs
+        question_mark!(
+            self.compute_meter
+                .consume(calc_bignum_cost!(self, bn_syscall_rhs.data_len as f64)),
+            result
+        );
+        // Get array pointer of bytes to consume
+        let rhs_data = question_mark!(
+            translate_slice::<u8>(
+                memory_mapping,
+                bn_syscall_rhs.data,
+                bn_syscall_rhs.data_len as u64,
+                self.loader_id,
+                true
+            ),
+            result
+        );
+        // Process BigNum call and populate results
+        let bn_syscall_result = question_mark!(
+            translate_type_mut::<FfiBigNumber>(
+                memory_mapping,
+                bn_ffi_out_addr,
+                self.loader_id,
+                true
+            ),
+            result
+        );
+        // Check max length of 1024 bytes
+        if bn_syscall_rhs.data_len > MAX_BIGNUM_BYTE_SIZE || bn_syscall_rhs.data_len < 1 {
+            *result = Err(SyscallError::BigNumberResultSizeError.into())
+        } else {
+            match BigNum::from_slice(rhs_data) {
+                Ok(bn_result) => bignum_to_ffibignumber(
```

Comment:
It does round-trip but a bad actor may attempt to put a slice bigger than 8192 b so the syscall adds the bounds checking

Reply:
If it round-trips, it's wasted cycles. Might as well copy it rather than processing through openssl, while possibly removing leading zeros as its big-endian. I can't see any advantage of doing the bounds check in a system call.

Reply:
Also adds access to function through "C" but @jackcmay advised bounds checks in syscalls...?

Reply:
Does OpenSSL do any other validation of the slice besides size?  If not, then this operation is probably pretty cheap and Sean is write that it doesn't need to be a syscall

Reply:
After more thought, maybe we want to keep the encoding of `value` away from the program.  

We have to do all security-related bounds checks in system calls because we cannot trust program space.  

@FrankC01, we should be doing these bounds check every time a syscall receives a FfiBigNumber because the program can pass whatever is wants

Reply:
@jackcmay So I check the result buffer size and, if passes that sniff test, I check that actual operation size. Are you saying every FfiBigNunber length should be checked for size coming into each syscall? I can do that, just want to confirm.

Reply:
Yeah, checking the result just makes sure the user has provided enough space but that should also be checked by the translate.  The real security risk is a user passing in some huge buffer and asking the runtime to crunch it, especially with a division.

Reply:
I can do that, I'm as concerned with `mul` and `mod_exp` as well...

Reply:
Add it to all the bignum syscall inputs